src/price_data.py

```diff
@@ -133,6 +133,91 @@ def _get_price_binance(
+        around 1 minute around `utc_time`.
+
+        None existing pairs like `TWTEUR` are calculated as
+        `TWTBTC * BTCEUR`.
+
+        Documentation: https://docs.pro.coinbase.com
+
+        Args:
+            start	        Start time in ISO 8601
+            end	            End time in ISO 8601
+            granularity	    Desired timeslice in seconds (60, 300, 900, 3600, 21600 or 86400)
+
+        Raises:
+            RuntimeError: Unable to retrieve price data.
+
+        Returns:
+            Decimal: Price of asset pair.
+        """
+
+        root_url = "https://api.pro.coinbase.com"
+        symbol = f"{base_asset}-{quote_asset}"
+        startTime = misc.to_iso_timestamp(utc_time - datetime.timedelta(minutes=5))
+        endTime = misc.to_iso_timestamp(utc_time + datetime.timedelta(minutes=5))
+        sub_url = f"/products/{symbol}/candles?start={startTime}&end={endTime}&granularity=60"
+        url = root_url + sub_url
+
+        log.debug("Calling %s", url)
+        response = requests.get(url)
+        data = json.loads(response.text)
+
+        # Some combinations do not exist (e.g. `TWTEUR`), but almost anything
+        # is paired with BTC. Calculate `TWTEUR` as `TWTBTC * BTCEUR`.
+        if isinstance(data, dict) and data.get("message") == "NotFound":
+            if quote_asset == "BTC" or base_asset == "BTC":
+                # If we are already comparing with BTC, we might have to swap
+                # the assets to generate the correct symbol.
+                # Check a last time, if we find the pair by changing the symbol
+                # order.
+                # If this does not help, we need to think of something else.
+                if swapped_symbols:
+                    raise RuntimeError(
+                        f"Can not retrieve {symbol=} from coinbase_pro")
+                # Changing the order of the assets require to invert the price.
+                log.debug("Getting price with swapped symbols...")
+                price = self.get_price(
+                    "coinbase_pro", quote_asset, utc_time, base_asset, swapped_symbols=True)
+                return misc.reciprocal(price)
+
+            btc = self.get_price("coinbase_pro", base_asset, utc_time, "BTC")
+            quote = self.get_price("coinbase_pro", "BTC", utc_time, quote_asset)
+            return btc * quote
+
+        response.raise_for_status()
+
+        if len(data) == 0:
+            log.warning("Coinbase Pro offers no price for `%s` at %s", symbol, utc_time)
+            return 0
+
+        # Find closest timestamp match
+        target_timestamp = misc.to_ms_timestamp(utc_time)
+        data_timestamps_ms = [int(float(d[2]) * 1000) for d in data]
+        closest_match_index = (
+            bisect.bisect_left(data_timestamps_ms, target_timestamp) - 1
+        )
+
+        # Use item in the middle if no closest match found
+        if closest_match_index == -1:
```

Comment:
The searched timestamp is older than all other timestamps when the function returns -1... It doesn't make sense to take the middle value.

Reply:
I'm not taking the middle value anymore, now that the implementation increases the time window if bisect couldn't find some item before and after the timestamp.